File: mongo_db/lib/models/task.dart
class Task {
  final String? id;
  String title;
  int priority;
  bool completed;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final String? status;

  Task({
    this.id,
    required this.title,
    required this.priority,
    this.completed = false,
    this.createdAt,
    this.updatedAt,
    this.status,
  });

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: json['_id'],
      title: json['title'] ?? '',
      priority: json['priority'] ?? 0,
      completed: json['completed'] ?? false,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
      status: json['status'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'title': title,
      'priority': priority,
      'completed': completed,
    };
  }
}

File: mongo_db/backend-api/models/Task.mjs
import mongoose from "mongoose";

const TaskSchema = new mongoose.Schema(
  {
    title: {
      type: String,
      required: [true, "Title is required"],
      trim: true,
      minlength: [1, "Title cannot be empty"],
      maxlength: [200, "Title cannot be longer than 200 characters"],
    },
    completed: {
      type: Boolean,
      default: false,
      index: true,
    },
    priority: {
      type: Number,
      default: 0,
      min: 0,
      max: 5,
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// Indexes
TaskSchema.index({ createdAt: -1 });
TaskSchema.index({ title: "text" });

// Virtual for task status
TaskSchema.virtual("status").get(function () {
  return this.completed ? "completed" : "pending";
});

// Pre-save middleware
TaskSchema.pre("save", function (next) {
  if (this.isModified("title")) {
    this.title = this.title.trim();
  }
  next();
});

const Task = mongoose.model("Task", TaskSchema);
export default Task;
